### ISSUE TITLE
setoid_rewrite fails on Equiv : Type -> Type -> Type

### ISSUE THREAD

**Issue by SimonBoulier** [coq/coq#6278]
The following setoid_rewrite fails with a universe error.
```coq
Require Import Classes.CMorphisms Classes.CRelationClasses.
Require Setoids.Setoid.

Axiom Equiv : Type -> Type -> Type.

Instance equivalence_equiv : CRelationClasses.Equivalence Equiv.
Admitted.

Instance prod_equiv : CMorphisms.Proper (Equiv ==> Equiv ==> Equiv) prod.
Admitted.

Lemma foo {A B C:Type} (X: Equiv A B) : Equiv (prod C A) (prod C B).
Proof.
  setoid_rewrite X.
```
The error is:
```
The 3rd term has type
 "crelation@{Top.57 Top.58} Type@{max(Top.22, Coq.Init.Datatypes.24)}"
which should be coercible to "crelation@{Top.83 Top.84} Type@{Top.1}".
```
The problem seems to be the type class search as the `eapply` works whereas nor `typeclasses_eauto` nor `eauto with typeclass_instances` does.

Changing `Equiv` to `Type -> Type -> Prop` does not solve the problem (but the error message is different) but changing it to `myType -> myType -> myType` with `myType := Type` monomorphic does.

This bug may also be related to https://github.com/coq/coq/issues/3254 but I'm not able to affirm it.

---

My goal is to use `setoid_rewrite` to rewrite equivalences in the HoTT library, as in https://github.com/HoTT/HoTT/issues/488. Do you think the implementation feasible?

EDIT now the error is "nothing to rewrite"

**Comment by Alizter** [comment]
This will be fixed by #14137

**Comment by SkySkimmer** [comment]
Still open / reopened by a parallel PR
